src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -213,11 +213,22 @@ private async Task RegisterDesignerAttributeAsync(Document document, string desi
             var updateService = await GetUpdateServiceIfCpsProjectAsync(document.Project, cancellationToken).ConfigureAwait(false);
             if (updateService != null)
             {
-                // we track work by async token but doesn't explicitly wait for it. and update service is free-thread service,
-                // no need to switch to UI thread to use it
-                var asyncToken = _listener.BeginAsyncOperation("RegisterDesignerAttribute");
-                _ = updateService.SetProjectItemDesignerTypeAsync(document.FilePath, designerAttributeArgument)
-                                 .ReportNonFatalErrorAsync().CompletesAsyncOperation(asyncToken);
+                try
+                {
+                    // we track work by async token but doesn't explicitly wait for it. and update service is free-thread service,
+                    // no need to switch to UI thread to use it
+                    var asyncToken = _listener.BeginAsyncOperation("RegisterDesignerAttribute");
+                    _ = updateService.SetProjectItemDesignerTypeAsync(document.FilePath, designerAttributeArgument)
+                                     .ReportNonFatalErrorAsync().CompletesAsyncOperation(asyncToken);
+                }
+                catch (ObjectDisposedException)
+                {
+                    // we might call update service after project is already removed and get object disposed exception.
+                    // we will catch the exception and ignore. this is wierd setup since we are catching exception from
+                    // async call, but until we have better way to handle this, this is what we have.
+                    // see this PR for more detail - https://github.com/dotnet/roslyn/pull/35383
+                    return;
```

Comment:
will this actually work?  is ObjectDisposedException thrown synchronously when SetProjectItemDesignerTypeAsync is called? or can it happen asynchronously?  If hte latter, should we not `await updateService.SetProjectItemDesignerTypeAsync`?

Reply:
@davkean  is SetProjectItemDesignerTypeAsync an actual `async` method?  If so, then this will presumably return a faulted task.  So i don't think this try/catch will suffice in the current form.

Reply:
I verified it throw exception synchronously and the callstack in the issue shows that it is thrown synchronously.

Reply:
In this _exact_ case, its thrown outside of async machinery, the method itself is not async.

Other things we're calling could also be disposed underneath of us inside of state machine (@Lifengl we should fix this - by wrapping in `LoadedProjectAsync<T>`) - so the safest mechanism as the code stands right now is to await and catch.

Reply:
> so the safest mechanism as the code stands right now is to await and catch.

That would be my preference.  it would be resilient to any future changes.  we can also then just do a try/finally/dispose on the BeginAsyncOperation bit.

Reply:
The code looks like this:

```` C#
       public Task SetProjectItemDesignerTypeAsync(string projectItemPath, string designerType)
        {
            return this.implementation.Value.SetProjectItemDesignerTypeAsync(projectItemPath, designerType); <!-- Accessing Value throws ObjectDisposedException
        }
````

Implementation looks like this:


``` C#
      public async Task SetProjectItemDesignerTypeAsync(string projectItemPath, string designerType)
      {
          // Call lots of things that might be disposed (CPS shouldn't let this happen(!))
      }
```
So we need to protect against both.

Reply:
can I see the code?

Reply:
> So we need to protect against both.

Makes sense.  let's use `await` then to make this clean and idiomatic and correct for all cases.

Reply:
like I said, @lifengl asked not to use await. it is weird fixing one might causing another issue such as it taking long time to run analysis since we have to wait for subtype changes. in that case, I am not sure why we even using new API rather than old one.

 if the SetProjectItemDesignerTypeAsync can throw in the middle of it running then really it is not a thread safe method, by the way.

anyway, I will make it to ignore disposable exception from continuation.

Reply:
by the way, won't my previous change better approach then? doing the change from UI thread? and checking heriarch still exist on UI thread?

Reply:
> anyway, I will make it to ignore disposable exception from continuation.

Sure.  Running from teh continuation would be fine for me.

> if the SetProjectItemDesignerTypeAsync can throw in the middle of it running then really it is not a thread safe method, by the way.

Agreed.  But i think we decided much earlier that this is an expedient fix for now while a better approach can be hashed out in the future.

Reply:
> by the way, won't my previous change better approach then? doing the change from UI thread? and checking heriarch still exist on UI thread?

i'd personally prefer to avoid the UI thread if we can avoid it.  IN the past, spamming the UI thread with lots of calls definitely could impact the user.  So if we can do this is a totally ui-less way that would be better IMO.

Reply:
ya, but we already do that for csproj and we know number of users of csproj is way way way higher than cps users. so not sure how much we actually buy from it except some theoretical thing.

Reply:
> i think we decided much earlier that this is an expedient fix for now while a better approach can be hashed out in the future.

what I meant by it is that I am trying to use it from BG with the belief that it is thread  safe API that I can call from BG, more and more it looks like that is not the case. so feels like it might just much simpler and safer to just call from UI thread.

Reply:
> In this exact case, its thrown outside of async machinery, the method itself is not async.
> Other things we're calling could also be disposed underneath of us inside of state machine (@lifengl we should fix this - by wrapping in LoadedProjectAsync<T>) - so the safest mechanism as the code stands right now is to await and catch.

I updated our inner implementation layer to use LoadedProjectAsync<>, and will still throw ObjectDisposedException, when the project is disposed at that point to match your changes.

Our outer implementation doesn't hold a reference to the project, so it cannot use LoadedProjectAsync.  We did this for most our COM layer objects, so if anyone holds it after the project is closed, we don't leak extra memory other than the shim object.

Reply:
why? if you call the service on the UI thread after a project is closed, it will throw in the exactly same way. The only thing different is that you know when the project is closed on the UI thread.  You do have a way to know when the project is closed on the UI thread, and based on that to manage lifetime of your own background tasks. I don't see anything new thing from the project system is necessary to address this issue.

Reply:
It's impossible write code here that doesn't throw ObjectDisposedException, unless he also blocks UI thread to avoid close.